src/assets/wise5/vle/timed-node/timed-node.component.ts

```diff
@@ -0,0 +1,191 @@
+        .map((event) => event.componentId);
+      this.advanceCurrentComponentIndex(componentsViewed);
+      this.setStepCompletedIfNecessary();
+    }
+  }
+
+  private advanceCurrentComponentIndex(componentsViewed: string[]): void {
+    this.node.components.forEach((component, componentIndex) => {
+      if (componentsViewed.includes(component.id) && componentIndex >= this.currentComponentIndex) {
+        this.currentComponentIndex = componentIndex + 1;
+      }
+    });
+  }
+
+  private setStepCompletedIfNecessary() {
+    if (this.currentComponentIndex >= this.node.components.length) {
+      this.stepCompleted = true;
+    }
+  }
+
+  private async startComponentTimer(): Promise<void> {
+    this.saveComponentViewedEvent();
+    this.timerCountDown = this.componentTimers[this.currentComponentIndex];
+
+    this.currentInterval = setInterval(() => {
+      if (--this.timerCountDown <= 0) {
+        this.componentCompleted();
+        clearInterval(this.currentInterval);
+      }
+    }, 1000);
+  }
+
+  private saveComponentViewedEvent() {
+    this.studentDataService.saveEvent(
+      'VLE',
+      this.node.id,
+      this.getCurrentComponentId(),
+      null,
+      'Timed',
+      'componentViewed',
+      {}
+    );
+  }
+
+  private getCurrentComponentId(): string {
+    return this.components.at(this.currentComponentIndex).id;
+  }
+
+  private getComponentSubmitArgs(componentId: string) {
+    return {
+      nodeId: this.node.id,
+      componentId: componentId
+    };
+  }
+
+  private componentCompleted(): void {
+    this.saveUnsavedWork();
+    this.hideComponents(++this.currentComponentIndex);
+    if (this.currentComponentIndex < this.node.components.length) {
+      this.startComponentTimer();
+    } else {
+      this.stepCompleted = true;
+      this.timedNodeService.broadcastIsNodeCompleted(true);
+    }
+  }
+
+  private saveUnsavedWork() {
+    if (!this.isPreview()) {
+      const currentComponentId = this.node.components[this.currentComponentIndex].id;
+      if (this.dirtyComponentIds.includes(currentComponentId)) {
+        this.studentDataService.broadcastComponentSubmitTriggered(
+          this.getComponentSubmitArgs(currentComponentId)
+        );
+      }
+    }
+  }
+
+  private hideComponents(showComponentIndex: number): void {
+    this.components.forEach((component, index) => {
+      this.componentToVisible[component.id] = index === showComponentIndex;
+    });
+  }
+
+  protected secondsToTimerDisplay(secondsLeft: number): string {
+    const minutes = '' + Math.floor(secondsLeft / 60);
+    const seconds = '' + (secondsLeft % 60);
+    return `${this.addZeroPadding(minutes)}${minutes}:${this.addZeroPadding(seconds)}${seconds}`;
+  }
+
+  private addZeroPadding(digits: string): string {
+    const numZeros = 2 - digits.length;
+    return '0'.repeat(numZeros > 0 ? numZeros : 0);
+  }
+
+  protected updateComponentVisibility(): void {
+    return;
```

Comment:
This function shouldn't return anything. Instead, it'd be good to have a comment about why we're overloading it.
```suggestion
    // do nothing because ...
```